162: preserve std::vector and std::unordered_map capacity

cz4rs opened the pull request:
fixes #162

cz4rs committed:
#162: container: preserve std::vector capacity

- serialize std::vector capacity and use reserve() to preserve it after
deserialization
- use reserve() before resize() to avoid reallocating twice
- test: check capacity after deserialization
- test: account for additional bytes serialized in traversal testa
- test: improve readability of traversal test

cz4rs commented:
note: `std::unordered_multimap` should probably get the same treatment

PhilMiller commented on src/checkpoint/container/vector_serialize.h:
```diff
@@ -60,6 +60,10 @@ typename std::enable_if_t<
+
+  SerialSizeType vec_capacity = vec.capacity();
+  s | vec_capacity;
+  vec.reserve(vec_capacity);
```
This should be done before resizing, so that it doesn't reallocate twice.

PhilMiller commented on tests/unit/test_container.cc:
```diff
@@ -97,6 +97,22 @@ static void testContainer(bool is_ordered, std::initializer_list<T> lst) {
+  auto t1 = checkpoint::deserialize<std::vector<int>>(ret->getBuffer());
+
+  EXPECT_EQ(c1.size(), t1->size());
+  EXPECT_EQ(c1.capacity(), t1->capacity());
```
The reserved capacity should be a named constant, and the comparison should be that the deserialized vector has capacity greater than or equal to both the constant and the original vector. The call to `reserve` can over allocate at both sites

cz4rs replied:
fixed, explanation added in the comment

PhilMiller commented:
The vector code looks good and as expected.

cz4rs committed:
#162: container: preserve std::unordered_map bucket count

- serialize std::unodered_map and std::unordered_multimap bucket count
to guarantee at least the same number of buckets after deserialization
- test: check bucket count after deserialization

PhilMiller commented:
Do the hash table types also need to do something to preserve maximum load factor or other parameters?

cz4rs commented:
> Do the hash table types also need to do something to preserve maximum load factor or other parameters?

good catch with max load factor, it should be (de)serialized before rehashing
other than that, I don't see anything in unordered_map interface that could be preserved

cz4rs committed:
#162: container: preserve maximum load factor for hash table types

- preserve bucket_count and max_load_factor for unordered associative
containers
- test: check max_load_factor after deserialization

lifflander reviewed (approved):
Looks good to me

cz4rs commented on tests/unit/test_container.cc:
```diff
@@ -82,28 +82,62 @@ static void testEqualityContainerUnordered(ContainerT& c1, ContainerT& t1) {
+  EXPECT_EQ(c1.size(), t1->size());
+  EXPECT_EQ(t1->max_load_factor(), custom_max_load_factor);
+  EXPECT_EQ(t1->max_load_factor(), c1.max_load_factor());
+  EXPECT_GE(t1->bucket_count(), custom_bucket_count);
```
I wasn't sure about `greater or equal` here (considering guaranteed low load factor), but test fails using `EXPECT_EQ`

cz4rs commented:
Thanks Jonathan! @PhilMiller if you have no more comments, I will merge this shortly
